### DISCUSSION TITLE
Option to set the start time of a rules query

### DISCUSSION THREAD

**Discussion by SysAdmLS** [jertel/elastalert2#806] [Q&A] [answered]
Hi, 
im indexing certificate expiry dates in elasticsearch using their expiration date as the timestamp. The problem now is that elastalert has no way (atleast i didn't find a way) to configure where to start the query. I can't say it should start the query from now+60d (since timeframe is always backwards).

Sample data:
Cert A : notvalidafter: 2023-03-02T23:59:59
Cert B: notvalidafter: 2024-02-02T23:59:59
Cert C: notvalidafter: 2025-03-18T23:59:59

Using elastalert-test-rule --start 2023-03-01  --end 2023-03-30 i was able to get a hit.  But there seems to be no way to control either the behaviour of timeframe nor to just let me control the time range i want to query with my rule (for example in the query itself).

Do u guys have any idea to make this work ?
Thanks in adavance

**Comment by jertel** [comment, marked as answer]
Have you tried using a negative `query_delay` value?

https://elastalert2.readthedocs.io/en/latest/ruletypes.html?highlight=delay#query-delay

**Reply by SysAdmLS** [reply]
Nice idea, wouldn't have had that idea myself.
That somewhat worked but doesn't work when i would add new found certificates in timeframes that elastalert already "scanned", correct ? 

The best solution for my would be to just control the timeframe myself in the query and get the matching results (that would also be way faster to query). But that seems to be against the way elastalert is designed to work if iam not mistaken?

The next best workaround is to use the --start parameter (--start <timestamp> will force ElastAlert 2 to begin querying from the given time, instead of the default, querying from the present.).

Or is there any way i could implement a dynamic query start and endtime with a custom ruletype ? (https://elastalert2.readthedocs.io/en/latest/recipes/adding_rules.html#tutorial) 

Thanks again for your help

**Reply by jertel** [reply]
Since you expect there to be new data showing up in already-scanned index time ranges I think the best option for your use case is to run in standalone, via the --start and --end args. Then use an external scheduler like cron to ensure elastalert starts up at least once a day.

**Reply by SysAdmLS** [reply]
I also thought about that but there is code preventing this in place.
https://github.com/jertel/elastalert2/blob/cf1ebb94b92e2d78961a449bb19eb171995b90b4/elastalert/elastalert.py#L838
WARNING:elastalert:Attempted to use query start time in the future (2023-02-01 00:00:00+00:00), sleeping instead

Also this will only work once because elastalert writes the last run endtime to the elasticsearch metadata.

**Reply by jertel** [reply]
Shouldn't the starting time remain the current time, and instead only adjust the ending time (via --end or `delay_query`), to find all pending expiration certs?

The `scan_entire_timeframe` may provide a means to workaround the other issue: https://elastalert2.readthedocs.io/en/latest/ruletypes.html?highlight=scan_entire_timeframe#scan-entire-timeframe

**Reply by SysAdmLS** [reply]
> Shouldn't the starting time remain the current time, and instead only adjust the ending time (via --end or `delay_query`), to find all pending expiration certs?

Yeah that makes sense.

> The `scan_entire_timeframe` may provide a means to workaround the other issue: https://elastalert2.readthedocs.io/en/latest/ruletypes.html?highlight=scan_entire_timeframe#scan-entire-timeframe
Yes although this messes with the --start time as u have to set a timeframe.

Im thinking about removing the future check for the starttime, do u see any problems regarding that ? This should probably fix my issues bc i can then just use a negative timeframe, for example days: -60 and use scan_entire_timeframe.

**Reply by jertel** [reply]
I don't have an issue with you submitting a PR that adds an option for future start times. But I don't like removing it altogether, to accommodate your unique use case, as it could cause problems for others that accidentally set it to the future and miss out on alerts without knowing.

Perhaps the new option could be called

`allow_future_timeranges` defaults to false

**Reply by SysAdmLS** [reply]
I didnt mean to remove it, I also can try implementing that option. I was just asking if you see any problems arising logic wise in other parts of elastalert if one does use a start time in the future.

**Reply by jertel** [reply]
No, I'm not aware of any potential problems.

**Reply by SysAdmLS** [reply]
I just tried to setup elastalert2 in my pycharm ide but i get weird import errors when i try to run elastalert.py 

Traceback (most recent call last):
  File "/home/itservice/PycharmProjects/elastalert2/elastalert/elastalert.py", line 32, in <module>
    from elastalert.alerters.debug import DebugAlerter
  File "/home/itservice/PycharmProjects/elastalert2/elastalert/elastalert.py", line 32, in <module>
    from elastalert.alerters.debug import DebugAlerter
ModuleNotFoundError: No module named 'elastalert.alerters'; 'elastalert' is not a package

Im unable to fix this, setting sources dir/working dir to elastalert2 project folder didn't help either. 
Im using python 3.10 and have already installed all the requirements but i cant get the imports to work.

**Reply by SysAdmLS** [reply]
Getting the same error using python 3.10 venv and vscode 1.66.1. Would appreciate if u could guide me into the right direction on how to setup a dev envoriment or what im doing wrong. @jertel

**Reply by SysAdmLS** [reply]
Ok, i figured it out myself i think. The runner has to be configured to execute the module (https://stackoverflow.com/a/51268846 using elastalert.elastalert) and not the elastalert.py script. This may be a point to include in the docs, on how to setup a dev env.

**Reply by nsano-rururu** [reply]
I am developing with venv and vscode, but there is no problem. elastic serach is executed first with docker. After modifying the code and executing the following command, run the rules of elastalert2 and check.

```
$ pip install "setuptools>=11.3"
$ python setup.py install
```

**Reply by SysAdmLS** [reply]
> I am developing with venv and vscode, but there is no problem. elastic serach is executed first with docker. After modifying the code and executing the following command, run the rules of elastalert2 and check.
> 
> ```
> $ pip install "setuptools>=11.3"
> $ python setup.py install
> ```

That seems like an inefficient way to debug and test source code changes because u would need to build elastalert everytime u change something and u can not use a debugger/breakpoints that way, right ?

I suggest u to try to create a launch.json for your vscode enviroment like explained here https://stackoverflow.com/questions/46102228/how-to-debug-a-python-module-in-visual-studio-codes-launch-json

Not tested but should work:
```json
    "configurations": [
    {
        "name": "Python: any name like script_file",
        "type": "python",
        "request": "launch",
        "module": "elastalert.elasalert"
    }
]
```

**Reply by nsano-rururu** [reply]
is that so. I think you can do whatever you want.

**Reply by nsano-rururu** [reply]
I don't know what I want to do

**Reply by SysAdmLS** [reply]
@jertel i just tried to implement the allow_future_timeranges option but there are many obstacles that im meeting.

Goal was to use a negative timeframe and therefore not querying in the past but the future, so endtime would be in the future. 

Problem with that is:
endtime is "hardcoded" in https://github.com/jertel/elastalert2/blob/cc3e5d0c3ad5789242f92cd8f27cdccd221e880d/elastalert/elastalert.py#L1200 to be ts_now() and cant be controlled from a rule.

and the starttime seems to be generated dynamically in https://github.com/jertel/elastalert2/blob/cc3e5d0c3ad5789242f92cd8f27cdccd221e880d/elastalert/elastalert.py#L661 using that endtime 

where btw im not really sure if timeframe is really evaluated unless used in combination with scan_entire_timeframe : True. That could also explain the issues described here https://github.com/jertel/elastalert2/discussions/815

My workaround for the endtime/starttime problem was to just swap the values but that is really dirty and messes with some logging later on plus only work when using scan_entire_timeframe because set_starttime doesn't seem to use the timeframe otherwise:
```python
# Don't run if starttime was set to the future
        if ts_now() <= rule['starttime'] and not rule['allow_future_timeranges'] :
            elastalert_logger.warning("Attempted to use query start time in the future (%s), sleeping instead" % (starttime))
            return 0
        # if allow_future_timeranges is set and the timeframe is negative
        # (self.set_starttime(rule, endtime) produces a starttime in the future)
        # so we swap endtime and starttime
        elif rule['allow_future_timeranges'] and rule['timeframe'] < datetime.timedelta(seconds=1) and rule.get :
            rule['starttime'], endtime = endtime, rule['starttime']
            rule['original_starttime'] = rule['starttime']```

**Reply by SysAdmLS** [reply]
My idea would be to make start and endtime controlable from a rule, for example:

starttime: now
endtime: now+15d

But that conflicts with many other parts in the code if I'am not mistaken.
And I dont really see how timeframe is acknowledged unless scan_entire_timeframe is set.

**Reply by SysAdmLS** [reply]
I adjusted my indexing code to include another timestamp of when the data was indexed, im then using this in the rule by declaring timestamp_field: timestamp. So i can use elastalert the way its designed to work while still being able to filter for expiring certificates in the future using filter: -range. 

I will therefore close this discussion.
